### PR TITLE
Add "delete" command

### LINKED ISSUE #501: delete command
By popular request of @satra and to reflect that web ui would never become as flexible as shell globbing, we need `dandi delete`.

- should be pretty much a trimmed down copy of `download` in terms of interface/handling inputs (paths or URLs) and just instead of `download` to `delete` of local and remote files
  - may be worth adding an option `--remote-only` to not bother with possibly present local files
- before jumping to the action, should ask user confirmation alike "Do you want to delete {len(assets)} from the {dandiset} on the server and {len(files)} local files?"  (later we might want to add a cmdline option to avoid this interaction, ok without for now since we better make it uniform with other commands like `upload` etc)
- tests should test against an instance of fixtured dandi-api that
  - deletion is forbidden by API server for assets in released versions
  - deletion in the draft does not delete assets (by downloading so we are certain that blobs are not removed from keystore) present in prior released versions
- no need for `girder` support - just raise NotImplementedError
 
Just ideas for later
- we might like to inform user on deletion of how many assets on remote server would cause those assets to no longer be present in the archive (i.e. not present in any other prior version of the dandiset)

### PR DESCRIPTION
**Pull request by jwodder**
Closes #501.

### CODE REVIEW

**Comment by jwodder** [comment]
@yarikoptic I'm unclear on how to properly invoke pyout.  My current attempt seems to result in pyout not consuming the iterators of progress dicts, and thus nothing is getting deleted.  (It works with `--devel-debug`, though.)

**Comment by jwodder** [comment]
@yarikoptic @dchiquito I'm writing tests for the `delete()` function, but when running the entire test suite, it keeps failing because Django is erroring out with "psycopg2.OperationalError: FATAL: sorry, too many clients already".  Putting various things in sessions doesn't help.

**Comment by yarikoptic** [comment]
@jwodder - is that for an asset delete or a dandiset delete?  smells like an issue to file against `dandi-api` since doesn't feel like anything to fix in dandi-cli.

**Comment by yarikoptic** [comment]
@jwodder also does it work when you try DELETE on the api.dandiarchive.org or specific to docker-compose instance?

**Comment by jwodder** [comment]
@yarikoptic I've only tested asset deletions so far, but the error is actually coming later in the test suite, after the deletion tests run — specifically, at `test_new_upload_extant_bad_existing` or adjacent.

**Comment by jwodder** [comment]
@yarikoptic I was able to get the tests to pass again by making most dandi-api tests use the same `Session` instance, but I'm not convinced we won't run into this problem again later.

**Comment by yarikoptic** [comment]
@jwodder please try it on the live instance.  Can do on any dandiset really besides that monster https://gui-beta-dandiarchive-org.netlify.app/#/dandiset/010000 which I am still uploading

**Comment by jwodder** [comment]
@yarikoptic Try what on the live instance, exactly?  I already tried deleting assets from the live instance, and it worked.

**Comment by yarikoptic** [comment]
I meant to try deleting (lots of?) assets on the live instance to see if similar "too many clients already" message would emerge in the dandi-api logs.

File: dandi/delete.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    version_id: str
+    asset_id: str
+    path: str
+
+
+@dataclass
+class Deleter:
+    client: Optional[DandiAPIClient] = None
+    dandiset_id: Optional[str] = None
+    deleting_dandiset: bool = False
```
Comment (yarikoptic) [inline dandi/delete.py:27]: please add a comment on what is the role of this `bool`

File: dandi/delete.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+class RemoteAsset(NamedTuple):
+    dandiset_id: str
+    version_id: str
+    asset_id: str
+    path: str
+
+
+@dataclass
+class Deleter:
```
Comment (yarikoptic) [inline dandi/delete.py:24]: please add a docstring describing the purpose for this class

File: dandi/delete.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            )
+        self.remote_assets.add(
+            RemoteAsset(dandiset_id, version_id, asset["asset_id"], asset["path"])
+        )
+
+    def register_asset_folder(
+        self, api_url: str, dandiset_id: str, version_id: str, folder_path: str
+    ) -> None:
+        if version_id != "draft":
+            raise ValueError("Cannot delete assets from published versions")
```
Comment (yarikoptic) [inline dandi/delete.py:73]: should we just rely on dandi-api to disallow such action and client to not care?  sorry if I forgot if we ran into some stumbling point on that or what -- if to keep, a comment should be added on why we need this client side logic

File: dandi/delete.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                asset_id["dandiset_id"],
+                asset_id.get("version") or "draft",
+                asset_id["location"],
+            )
+        else:
+            raise RuntimeError(f"Unexpected asset type for {url}: {asset_type}")
+
+    def register_local_path_equivalent(self, instance_name: str, filepath: str) -> None:
+        instance = known_instances[instance_name]
+        if instance.metadata_version == 0:
```
Comment (yarikoptic) [inline dandi/delete.py:116]:
```suggestion
        if instance.girder:
```
to make it more explicit than some possible relationship between having metadata version and girder

File: dandi/delete.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+            from .support import pyout as pyouts
+
+            pyout_style = pyouts.get_style(hide_if_missing=False)
+            rec_fields = ("path", "status", "message")
+            out = pyouts.LogSafeTabular(
+                style=pyout_style, columns=rec_fields, max_workers=jobs
+            )
+            with out:
+                for gen in deleter.process_assets_pyout():
+                    out(gen)
```
Comment (yarikoptic) [inline dandi/delete.py:199]: please add at least a basic smoke test run with pyout as a backend to ensure that  no gotchas with pyout rendering etc,so we do not miss some breakage

**Review by yarikoptic — CHANGES_REQUESTED** [review]
Looks great, thank you @jwodder .  I left just minor wishlists, the primary of which is to add a test run with pyout so we do not miss/spill some obvious errors into user faces.

**Comment by jwodder** [comment]
@yarikoptic Review comments addressed.

**Comment by jwodder** [comment]
@yarikoptic I deleted all 1328 assets in Dandiset 000008 with a single command, and there were no errors.

**Comment by yarikoptic** [comment]
Thank you @jwodder !

**Pull request merged**